### ISSUE TITLE
import status for existing tiddlers

### ISSUE THREAD

**Issue by tobibeer** [Jermolene/TiddlyWiki5#2240]
Upon importing any tiddlers — the import of which is not otherwise blocked by TiddlyWiki — the status column must clearly indicate that a given tiddler already exists in the target wiki. Preferably, the whole line will have a different highlight, e.g. be red.

An existing tiddler to be imported should be deselected by default, so a user would have to actively select / toggle the checkbox for each (or possibly all tiddlers at once) and then import.

**Comment by gesundelehre** [comment]
+1 for me

It would also be helpful, if one could then compare the existing tiddler with the one that is to be imported

**Comment by danielo515** [comment]
I wrote a plugin for that functionality. It was a long time ago, so I'm not sure if it stills working with the latest version, but what it does is unchecking the import checkbox for those tiddlers that already exists. In fact, I discovered a small problem while developing it, there is an issue about it #758. When you uncheck a tiddler progragmatically it's content is destroyed, which is not very good if you want to import it after checking.

I faced this problem several times, when I'm not sure what version of my personal wiki is the most recent. I would love to just drop the newest on top of the oldest, but then I'm on risk of overwriting tiddlers.

**Comment by tobibeer** [comment]
added 

> An existing tiddler to be imported should be deselected by default, so a user would have to actively select / toggle the checkbox for each (or possibly all tiddlers at once) and then import.

**Comment by Jermolene** [comment]
Hi @tobibeer 

> Upon importing any tiddlers — the import of which is not otherwise blocked by TiddlyWiki — the status column must clearly indicate that a given tiddler already exists in the target wiki. Preferably, the whole line will have a different highlight, e.g. be red.

My own preference is to have an expandable diff of the differences between the incoming tiddler and the existing tiddler.

> An existing tiddler to be imported should be deselected by default, so a user would have to actively select / toggle the checkbox for each (or possibly all tiddlers at once) and then import

That doesn't seem right to me, can you expand on the reasoning? Isn't clicking the "import" button sufficient confirmation of the users desire to re-import the same tiddler?

**Comment by danielo515** [comment]
> My own preference is to have an expandable diff of the differences between the incoming tiddler and the existing tiddler.

That is indeed a cool idea. But if we put such high expectations the basic feature will not be implemented.

>That doesn't seem right to me, can you expand on the reasoning? Isn't clicking the "import" button sufficient confirmation of the users desire to re-import the same tiddler?

It is more frequent that you don't want to overwrite existing tiddlers than the opposite. If you are importing an older version of your own wiki (because maybe you made some editions on it by mistake) you usually only want the different tiddlers. Being forced to uncheck one hundred boxes because you only want to get the three that are missing is a tedious task. 

Also a "select all" and "unselect all" button is very needed because what I said before. You could unselect all and then select specifically those that you want and the opposite. This is nothing strange or custom, many many user interfaces that have switches implement this feature.

**Comment by telmiger** [comment]
Hi all! For the case @danielo515 described above

> when I'm not sure what version of my personal wiki is the most recent.

it would be helpful just to see which version is newer, e.g. a warning, that the tiddler on the import list is an "older version" refering to the modified timestamp. (This could take some pressure from the diff functionality.)

As for the default checking I would go with @Jermolene – when a tiddler shows up on my import list, this is a clear indication that I actually intend to import it. (When I want to restore an old version of a tiddler, I drag a link to my wiki, get a confirmation that it actually is an older version and confirm the import. Having to re-check the checkbox would be one step too much.)

**Comment by tobibeer** [comment]
I think what it really boils down is the right UX to allow the user to make the decision he's after, could be to overwrite but could just as well be not to. so, we need options to make pre-selections:

* select all
* select "safe" ...meaning: only those that would not overwrite stuff

**Comment by Jermolene** [comment]
The difficulties of the checkboxes is covered in another ticket, I believe. To focus on the OP, perhaps what you want is for incoming tiddlers to be unchecked by default if (a) they exist in the current wiki and (b) they are older than the tiddler in the current wiki. There would need to be some more detailed rules covering the situation of a missing `modified`/`created` field.

**Comment by pmario** [comment]
>To focus on the OP, perhaps what you want is for incoming tiddlers to be unchecked by default if 

> (a) they exist in the current wiki and 
 (b) they are older than the tiddler in the current wiki. 
There would need to be some more detailed rules covering the situation of a missing modified/created field.

I'd like to add: 

- force import of **status** tiddlers, to allow tiddlers.json to be prepared for users. So they can import the file and get an exact state, that eg: opens tiddler info area, and / or selects certain tabs in the ControlPanel ... 

This would be extremely valuable for documentation purpose.

**Comment by pmario** [comment]
Probably there should be an **force import blocked plugins** too, for users, that know what they do.

**Comment by danielo515** [comment]
Regarding what Mario said, there is an issue already open #758. Blocked tiddlers are destructively discarded

**Comment by twMat** [comment]
Bump. This is a very reasonable and universal proposition.

**Comment by saqimtiaz** [comment]
> Upon importing any tiddlers — the import of which is not otherwise blocked by TiddlyWiki — the status column must clearly indicate that a given tiddler already exists in the target wiki. Preferably, the whole line will have a different highlight, e.g. be red.

This is now implemented for 5.1.23

> 
> An existing tiddler to be imported should be deselected by default, so a user would have to actively select / toggle the checkbox for each (or possibly all tiddlers at once) and then import.

This as well as Jeremy's idea of showing a diff of the incoming tiddler vs the existing ones are worth pursuing going forwards.

**Comment by Jermolene** [comment]
> This as well as Jeremy's idea of showing a diff of the incoming tiddler vs the existing ones are worth pursuing going forwards.

That was done, and can be accessed from the dropdown at the top of the import tiddler.

**Comment by saqimtiaz** [comment]
@Jermolene right, thank you for pointing that out. 

So the only outstanding item is not having existing tiddlers selected by default if they are older than existing tiddler.

**Comment by Jermolene** [comment]
The other thing I've often wanted is to somehow hide incoming tiddlers that are identical to the existing ones. Perhaps we'd need a checkbox and show them by default because there is some potential for confusion

**Comment by pmario** [comment]
@Jermolene ... IMO this issue can be closed. We do have a different colour now and there is a possibility to rename tiddlers at import

**Comment by Jermolene** [comment]
Thanks @pmario
